src/skeptic/lib.rs

```diff
@@ -166,22 +166,35 @@ enum Buffer {
 }
 
 fn extract_tests_from_file(path: &Path) -> Result<DocTest, IoError> {
-    let mut tests = Vec::new();
-    // Oh this isn't actually a test but a legacy template
-    let mut old_template = None;
-
     let mut file = try!(File::open(path));
     let ref mut s = String::new();
     try!(file.read_to_string(s));
-    let mut parser = Parser::new(s);
-
-    let mut buffer = Buffer::None;
 
     let ref file_stem = sanitize_test_name(path.file_stem().unwrap().to_str().unwrap());
-    let mut section = None;
+    // Oh this isn't actually a test but a legacy template
+    let mut old_template = None;
 
     // In order to call get_offset() on the parser,
     // this loop must not hold an exclusive reference to the parser.
+    let tests = extract_tests_from_string(s, file_stem, &mut old_template);
+
+    let templates = load_templates(path)?;
+
+    Ok(DocTest {
+        path: path.to_owned(),
+        old_template: old_template,
+        tests: tests,
+        templates: templates,
+    })
+}
+
+fn extract_tests_from_string(s: &mut String, file_stem: &String, old_template: &mut Option<String>) -> Vec<Test> {
```

Comment:
`s: &mut String` does not need to be `&mut`. The bare `&` will do here.
Also using `&mut` as "pseudo return" is not the best solution. It would be much better to just return it along the Vec in a tuple ie:
```rust
fn extract_tests_from_string(s: &String, file_stem: &String) -> (Option<String>, Vec<Test>)
```

Reply:
I agree, output variables are a code smell, I'm torn between my desire to take the time required to  write tested code and leaving something that can be refactored and tested later by someone else.... gah balancing open source with life commitments is hard :s

Reply:
No problem mate. You have already done a lot of valuable refactoring and we are on a finish line with this PR. No need to rush things :)